File: JIRA/sync-jira.js
#!/usr/bin/env node

require('dotenv').config();
const axios = require('axios');
const fs = require('fs');
const path = require('path');

const JIRA_EMAIL = process.env.JIRA_EMAIL;
const JIRA_API_TOKEN = process.env.JIRA_API_TOKEN;
const JIRA_DOMAIN = process.env.JIRA_DOMAIN;
const JIRA_PROJECT = process.env.JIRA_PROJECT;

if (!JIRA_EMAIL || !JIRA_API_TOKEN || !JIRA_DOMAIN || !JIRA_PROJECT) {
  console.error('❌ Erro: Faltam variáveis .env');
  process.exit(1);
}

const auth = {
  username: JIRA_EMAIL,
  password: JIRA_API_TOKEN
};

async function testAuth() {
  try {
    console.log('🔐 Testando autenticação...');
    const response = await axios.get(`${JIRA_DOMAIN}/rest/api/3/myself`, { auth });
    console.log(`✅ Autenticado como: ${response.data.displayName}\n`);
    return true;
  } catch (error) {
    console.error(`❌ Erro de autenticação: ${error.response?.status}`);
    return false;
  }
}

async function findBoardByProject() {
  try {
    const response = await axios.get(`${JIRA_DOMAIN}/rest/agile/1.0/board`, { auth });
    const boards = response.data.values || [];
    
    const board = boards.find(b => b.name.toUpperCase().includes(JIRA_PROJECT));
    if (board) {
      console.log(`📌 Board encontrado: [${board.id}] ${board.name}`);
      return board.id;
    }
    
    if (boards.length > 0) {
      console.log(`⚠️ Projeto "${JIRA_PROJECT}" não encontrado, usando primeiro board: ${boards[0].name}`);
      return boards[0].id;
    }
    
    return null;
  } catch (error) {
    console.error(`❌ Erro ao buscar boards: ${error.message}`);
    return null;
  }
}

async function getIssues(boardId) {
  try {
    console.log(`📡 Buscando issues do board ${boardId}...`);
    
    const response = await axios.get(
      `${JIRA_DOMAIN}/rest/agile/1.0/board/${boardId}/issue`,
      { auth, params: { maxResults: 100 } }
    );

    const issues = response.data.issues || [];
    console.log(`✅ ${issues.length} issues encontradas\n`);
    return issues;
  } catch (error) {
    console.error(`❌ Erro ao buscar issues: ${error.message}`);
    return [];
  }
}

function formatMarkdown(issues) {
  let md = `# 📋 JIRA Tasks - ${JIRA_PROJECT}\n\n`;
  md += `*Sincronizado em: ${new Date().toLocaleString('pt-PT')}*\n\n`;

  if (issues.length === 0) {
    md += 'Nenhuma tarefa encontrada.\n';
    return md;
  }

  const byStatus = {};
  issues.forEach(issue => {
    const status = issue.fields.status?.name || 'Unknown';
    if (!byStatus[status]) byStatus[status] = [];
    byStatus[status].push(issue);
  });

  const statusOrder = ['To Do', 'In Progress', 'In Review', 'Done'];
  
  statusOrder.forEach(status => {
    if (byStatus[status]) {
      md += `## ${status}\n\n`;
      byStatus[status].forEach(issue => {
        md += `- **[${issue.key}]** ${issue.fields.summary}\n`;
        md += `  - Assignee: ${issue.fields.assignee?.displayName || 'Unassigned'}\n`;
        md += `  - Priority: ${issue.fields.priority?.name || 'N/A'}\n`;
        md += `  - Link: [${JIRA_DOMAIN}/browse/${issue.key}](${JIRA_DOMAIN}/browse/${issue.key})\n\n`;
      });
    }
  });

  Object.keys(byStatus).forEach(status => {
    if (!statusOrder.includes(status)) {
      md += `## ${status}\n\n`;
      byStatus[status].forEach(issue => {
        md += `- **[${issue.key}]** ${issue.fields.summary}\n`;
        md += `  - Assignee: ${issue.fields.assignee?.displayName || 'Unassigned'}\n`;
        md += `  - Priority: ${issue.fields.priority?.name || 'N/A'}\n`;
        md += `  - Link: [${JIRA_DOMAIN}/browse/${issue.key}](${JIRA_DOMAIN}/browse/${issue.key})\n\n`;
      });
    }
  });

  return md;
}

async function main() {
  console.log('🚀 JIRA Sync - KAN Board\n');

  const authOk = await testAuth();
  if (!authOk) return;

  const boardId = await findBoardByProject();
  if (!boardId) {
    console.error('❌ Nenhum board encontrado');
    return;
  }

  const issues = await getIssues(boardId);
  if (issues.length === 0) {
    console.log('⚠️ Nenhuma tarefa para sincronizar.');
    return;
  }

  const markdown = formatMarkdown(issues);
  const outputFile = path.join(__dirname, `JIRA_TASKS_${JIRA_PROJECT}.md`);
  
  fs.writeFileSync(outputFile, markdown, 'utf-8');
  console.log(`✅ Tarefas salvas em: ${outputFile}`);
  console.log(`📊 Total: ${issues.length} issues\n`);
}

main();

File: JIRA/sync_jira.py
#!/usr/bin/env python3
"""
Sync JIRA tasks with local project
Pulls all issues from JIRA board and creates local TODO file
"""

import os
import json
import requests
from dotenv import load_dotenv
from datetime import datetime
from pathlib import Path

# Load environment variables from .env
load_dotenv()

JIRA_EMAIL = os.getenv('JIRA_EMAIL')
JIRA_API_TOKEN = os.getenv('JIRA_API_TOKEN')
JIRA_DOMAIN = os.getenv('JIRA_DOMAIN')
JIRA_PROJECT = os.getenv('JIRA_PROJECT')

if not all([JIRA_EMAIL, JIRA_API_TOKEN, JIRA_DOMAIN, JIRA_PROJECT]):
    print("❌ Erro: Faltam variáveis .env")
    print("Certifica-te que .env tem: JIRA_EMAIL, JIRA_API_TOKEN, JIRA_DOMAIN, JIRA_PROJECT")
    exit(1)

# JIRA API base URL
JIRA_API_URL = f"{JIRA_DOMAIN}/rest/api/3"

def get_jira_issues():
    """Fetch all issues from JIRA project"""
    url = f"{JIRA_API_URL}/projects/{JIRA_PROJECT}/issues"
    
    headers = {
        "Authorization": f"Basic {JIRA_EMAIL}:{JIRA_API_TOKEN}",
        "Content-Type": "application/json"
    }
    
    print(f"📡 Conectando a JIRA: {JIRA_DOMAIN}")
    
    try:
        # Use JQL to get all issues
        jql_url = f"{JIRA_API_URL}/search"
        params = {
            "jql": f"project = {JIRA_PROJECT}",
            "maxResults": 100,
            "fields": "key,summary,status,assignee,priority,created,updated"
        }
        
        response = requests.get(jql_url, headers=headers, params=params, auth=(JIRA_EMAIL, JIRA_API_TOKEN))
        response.raise_for_status()
        
        data = response.json()
        print(f"✅ Sincronizado: {len(data['issues'])} tarefas encontradas")
        return data['issues']
        
    except requests.exceptions.ConnectionError:
        print("❌ Erro: Não consegui conectar a JIRA. Verifica o JIRA_DOMAIN.")
        return []
    except requests.exceptions.HTTPError as e:
        print(f"❌ Erro HTTP: {e.response.status_code} - {e.response.text}")
        return []
    except Exception as e:
        print(f"❌ Erro: {str(e)}")
        return []

def format_issues_to_markdown(issues):
    """Convert JIRA issues to Markdown format"""
    if not issues:
        return "# 📋 JIRA Tasks\n\nNenhuma tarefa encontrada."
    
    md = f"# 📋 JIRA Tasks - {JIRA_PROJECT}\n\n"
    md += f"*Sincronizado em: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}*\n\n"
    
    # Group by status
    by_status = {}
    for issue in issues:
        status = issue['fields'].get('status', {}).get('name', 'Unknown')
        if status not in by_status:
            by_status[status] = []
        by_status[status].append(issue)
    
    # Sort statuses: To Do, In Progress, Done
    status_order = ['To Do', 'In Progress', 'Done']
    
    for status in status_order:
        if status in by_status:
            md += f"## {status}\n\n"
            for issue in by_status[status]:
                key = issue['key']
                summary = issue['fields'].get('summary', 'N/A')
                assignee = issue['fields'].get('assignee', {})
                assignee_name = assignee.get('displayName', 'Unassigned') if assignee else 'Unassigned'
                priority = issue['fields'].get('priority', {}).get('name', 'N/A')
                
                md += f"- **[{key}]** {summary}\n"
                md += f"  - Assignee: {assignee_name}\n"
                md += f"  - Priority: {priority}\n"
                md += f"  - Link: {JIRA_DOMAIN}/browse/{key}\n\n"
    
    # Add any other statuses not in the standard list
    for status in by_status:
        if status not in status_order:
            md += f"## {status}\n\n"
            for issue in by_status[status]:
                key = issue['key']
                summary = issue['fields'].get('summary', 'N/A')
                assignee = issue['fields'].get('assignee', {})
                assignee_name = assignee.get('displayName', 'Unassigned') if assignee else 'Unassigned'
                priority = issue['fields'].get('priority', {}).get('name', 'N/A')
                
                md += f"- **[{key}]** {summary}\n"
                md += f"  - Assignee: {assignee_name}\n"
                md += f"  - Priority: {priority}\n"
                md += f"  - Link: {JIRA_DOMAIN}/browse/{key}\n\n"
    
    return md

def main():
    """Main sync function"""
    print("🚀 Iniciando sincronização JIRA...\n")
    
    # Fetch issues
    issues = get_jira_issues()
    
    if not issues:
        print("⚠️ Nenhuma tarefa para sincronizar.")
        return
    
    # Format to Markdown
    markdown_content = format_issues_to_markdown(issues)
    
    # Write to file
    output_file = Path(__file__).parent / f"JIRA_TASKS_{JIRA_PROJECT}.md"
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(markdown_content)
    
    print(f"✅ Tarefas salvas em: {output_file}")
    print(f"📊 Total: {len(issues)} tarefas")
    
    # Print summary
    print("\n" + "="*50)
    print(markdown_content[:500] + "...")

if __name__ == "__main__":
    main()
